data/org.freedesktop.portal.CaptureInput.xml

```diff
@@ -0,0 +1,406 @@
+
+        Supported keys in the @options vardict include:
+        <variablelist>
+          <varlistentry>
+            <term>handle_token s</term>
+            <listitem><para>
+              A string that will be used as the last element of the @handle. Must be a valid
+              object path element. See the #org.freedesktop.portal.Request documentation for
+              more information about the @handle.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+
+        Supported keys in the @barriers vardicts include:
+        <variablelist>
+          <varlistentry>
+            <term>position iiii</term>
+            <listitem><para>
+              The x1/y1 x2/y2 position of the pointer barrier. A horizontal
+              pointer barrier must have y1 == y2, a vertical pointer barrier
+              must have x1 == x2. Diagonal pointer barriers are not supported.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>direction i</term>
+            <listitem><para>
+                The directional movement to trigger the pointer barrier. A
+                negative integer implies the pointer barrier triggers on negative
+                movement across the pointer barrier (e.g. "up"
+                for a horizontal barrier), a positive integer implies the
+                pointer barrier triggers on positive movement across
+                the pointer barrier (e.g. "right" for a vertical barrier).
+                A direction of zero triggers the barrier in either direction.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+
+        The following results get returned via the #org.freedesktop.portal.Request::Response signal:
+        <variablelist>
+          <varlistentry>
+            <term>failed_barriers au</term>
+            <listitem><para>
+              An array of indices of pointer barriers that have been denied. The
+              indices match the indices of the entries in the @barriers argument.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+    -->
+    <method name="SetPointerBarriers">
+      <arg type="o" name="session_handle" direction="in"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="aa{sv}" name="barriers" direction="in"/>
+      <arg type="o" name="handle" direction="out"/>
+    </method>
+
+    <!--
+        Enable:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @options: Vardict with optional further information
+
+        Enable input capturing. This does not immediately trigger capture, it
+        merely enables the capturing to be triggered at some future point
+        (e.g. by the cursor moving across a barrier). If and when that happens,
+        the Request object triggers the #org.freedesktop.portal.Request::Response
+        signal.
```

Comment:
Do we really want to communicate the `enabled` signal via a one shot request object? That would make moving the pointer back and forth across a trigger barrier racy, as one might try to attempt to move the pointer back into the intended barrier before the new `Enable()` was called.

Reply:
What I *want* to do is simply make this a signal on the session object like #713 :smile: And everything else, come to think of it. Failing that, keeping the request open until it triggered seemed like the best approach given the current infrastructure. More documentation is needed in that cancelling that request is akin to `Disable()` but otherwise it works all fine in my head.

I'm not sure on the raciness:
- application calls `Enable()` with the `Request` object
- pointer moves across the barrier, compositor starts re-routing events
- events start flowing on libei | `Request` object succeeds (in either order, can't do much about that bit)
- application calls `Disable()` | compositor decides its enough (in either order)
- compositor stops capturing, `Disabled` signal

Because it's input capture, there's no "moving back" once the compositor triggered, the input events are re-routed and the cursor doesn't move. This is a bit hairier when you're just capturing keyboard events (or a subset of pointer devices) based on pointer motion but even then the above works - once capturing starts, the event flow starts.

Reply:
> What I want to do is simply make this a signal on the session object like https://github.com/flatpak/xdg-desktop-portal/issues/713 smile

The signal could have the session handle as a parameter.

Anyhow, what I was thinking of is that when the capturing is disabled, there will be an amount of time where a "recapture" isn't possible because it hasn't been explicitly re-enabled. This can be avoided by adding a `Enabled` signal, and making `Enable()` create a session that can be enabled/disabled any amount of times. Maybe it'd be more clear what I was thinking if one renames things:
```
<method name="Enable">...</method>
<method name="Disable">...</method>
<method name="Release">...<arg name="position" type="(ii)"/></method>
<signal name="Activated">...</signal>
<signal name="Deactivated">...</signal>
```
So the flow would be

1. CreatePointerConstraints
1. Enable
...
1. Activated (events coming)
1. Release(x,y)
1. Deactivated (no more events)
...
1. Activated (events coming)
1. Deactivated (no more events)

Looking at that, the API becomes bigger, so maybe that's not good. What raises my eye brows is that I'm not sure requests have been used in this way before; they tend to be used to relatively short lived and represent the user making a decision, not a async "enable" request that might be responded to 2 hours later. As in, the request to enable enable the capture was already granted more or less immediately, it's just that it hasn't triggered yet, which makes it feel a bit odd.

Reply:
I wondered if a "oneshot" or "re-arm" arguments could help here. But no matter how I type it out, it comes down to the same thing: we need `Enable()` and `Disable()` and a `Disabled` signal. And we do need the `Activated/Deactivated` signals, I can't find a way to make the API smaller. So despite several attempts, I'm coming up with the same flow as you did above.

If (mis)using the Request like I did above is not ideal, I'm happy to change it to the "normal" flow that relies on signals.

Reply:
My gut feeling is that `Activated/Deactivated` is better than a very long lived request object. It'd make this API less "unique" which might be a good thing.

Reply:
done locally now, will be in the next push